pysmt/simplifier.py

```diff
@@ -280,47 +280,37 @@ def walk_exists(self, formula, args, **kwargs):
         return self.manager.Exists(varset, sf)
 
     def walk_plus(self, formula, args, **kwargs):
-        is_real = any(x.is_real_constant() for x in args)
-        is_int = any(x.is_int_constant() for x in args)
-        assert not (is_real and is_int)
-
-        if all(x.is_constant() for x in args):
-            res = sum(x.constant_value() for x in args)
-            if is_real:
-                return self.manager.Real(res)
+        new_args = []
+        constant_add = 0
+        stack = list(args)
+        ttype = self.env.stc.get_type(args[0])
+        is_algebraic = False
+        while len(stack) > 0:
+            x = stack.pop()
+            if x.is_constant():
+                if x.is_algebraic_constant():
+                    is_algebraic = True
+                constant_add += x.constant_value()
+            elif x.is_plus():
+                stack += x.args()
             else:
-                return self.manager.Int(res)
-
-        ns = [x for x in args if not x.is_constant()]
-        if len(ns) < len(args):
-            num = sum(x.constant_value() for x in args if x.is_constant())
-            if num != 0:
-                if is_real:
-                    ns.append(self.manager.Real(num))
-                else:
-                    ns.append(self.manager.Int(num))
-
-        if len(ns) == 2:
```

Comment:
To check: I think these cases would not be handled, because the Times subexpression would just endup in `newargs` but never simplified. However, since we simplify bottom up, it might be that our `walk_times` got smarter since we wrote this and this is actually dead code.

Reply:
Yes, you are right. I will re-add such rewriting.

In addition, using a simple dictionary it could be possible to perform some other simplifications like adding the constant coefficients of the terms of the addition having the same symbolic factors (`2*x*y + 3*x*y` into `5xy`).

Reply:
The general approach looks reasonable. However, the simplifier is called in several places, so we need to make sure that this is not significantly slower than it was before. We have some scripts in the root directory that we used to benchmark the parser. The same could be extended to benchmark the simplifier and show that these changes are not significantly slower.
https://github.com/pysmt/pysmt/blob/master/SMT-LIB/parse_all.py

It would be best to move those changes to a different PR in the interest of merging this PR sooner.

Reply:
Ok, I reverted the changes that implemented the simplification of linear combination of symbols. They have been stored on an other branch.

The current implementation implements a slight generalization of the previous simplification. All terms with negative constant are grouped in the rhs of a Minus expression.   
This implementation never goes down the AST more than a couple of levels.

There is a bit of code duplication to handle the constant factor of Times expressions.

I haven't looked at the benchmarking part yet.